README.md
## Null Safety support
Fork packages [flutter_aes_ecb_pkcs5](https://pub.dartlang.org/packages/flutter_aes_ecb_pkcs5) 

Solve the null safety support problem


## AES 128 pkcs5 only support
* Provide generation of AES keys
* Provide encrypt for AES keys  
* Provide decrypt for AES keys  

* Android uses java, iOS uses OC development


## Usage

Use this package as a library
1. Depend on it
Add this to your package's pubspec.yaml file:

```
dependencies:
  aes_ecb_pkcs5_flutter: ^0.1.1
```

2. Install it You can install packages from the command line:

with Flutter:

```
$ flutter pub get
```

Alternatively, your editor might support flutter pub get. Check the docs for your editor to learn more.

3. Import it Now in your Dart code, you can use:

```
import 'package:aes_ecb_pkcs5_flutter/aes_ecb_pkcs5_flutter.dart';
```



#### AES Use Example



```dart
import 'package:aes_ecb_pkcs5_flutter/aes_ecb_pkcs5_flutter.dart';

void main() {

    var data = "{\"username\":\"helloword\"}";
    
    //generate a 16-byte random key
    var key = await FlutterAesEcbPkcs5.generateDesKey(128);
    
    print(key);
    //encrypt
    var encryptText = await FlutterAesEcbPkcs5.encryptString(data, key);
    
    print(encryptText);
    //decrypt
    var decryptText  = await FlutterAesEcbPkcs5.decryptString(encryptText, key);
    
    print(decryptText);
  
}
```

#### Run Example Result
```
data:{"username":"helloword"}
create key:DA8FD2693B652D587EEE51E7ADD9DBB8
encryptText :851544E224F6C1C7F560AB82889DE0DF66EEB6762AB836F4EE13E31100F0EA16
decryptText :{"username":"helloword"}
```

#### License

The MIT License(http://opensource.org/licenses/MIT)

Please feel free to use and contribute to the development.




CHANGELOG.md
## 0.0.1

* first create aes ecb pkcs5 .

## 0.0.2

* update readme and Code comment.

## 0.1.0
* update readme and Code comment and description.
* its release.

## 0.1.1
* change description.

## 0.1.2
* support null-safety.
* support android plugin V2.


test/aes_ecb_pkcs5_flutter.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:aes_ecb_pkcs5_flutter/aes_ecb_pkcs5_flutter.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_aes_ecb_pkcs5');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await FlutterAesEcbPkcs5.platformVersion, '42');
  });
}


android/src/main/java/com/smallbuer/flutter_aes_ecb_pkcs5/AesSecurity.java
package com.smallbuer.flutter_aes_ecb_pkcs5;

import java.security.NoSuchAlgorithmException;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

/**
 * aes加密算法
 * AES/ECB/PKCS5Padding
 */
public class AesSecurity {

    private static final String ALGORITHM = "AES";
    private static final String TRANSFORMATION = "AES/ECB/NoPadding";
    private static final String DEFUALT_ENCODING = "UTF8";


    public  String generateDesKey(int length)  {
        //实例化
        KeyGenerator kgen = null;
        try {
            kgen = KeyGenerator.getInstance("AES");
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        //设置密钥长度
        kgen.init(length);
        //生成密钥
        SecretKey skey = kgen.generateKey();
        //返回密钥的二进制编码
        return Hex.encode(skey.getEncoded());
    }


    /**
     * 加密
     * @param  input 加密的字符串
     * @param  key   解密的key ,十六进制表示
     * @return HexString
     */
     
    public String encrypt(String input, String keyHex){

        byte[] encryptKey = Hex.decode(keyHex);
        byte[] encryptedByteArray = null;
        try {
//            if (command.length < 16) {
//                throw new Exception("Command must be 16 bytes");
//            }
            byte[] keyByteArray = new byte[32];
            System.arraycopy(encryptKey, 0, keyByteArray, 0, 32);
            SecretKeySpec key = new SecretKeySpec(keyByteArray, "AES");
            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
            cipher.init(Cipher.ENCRYPT_MODE, key);
            encryptedByteArray = cipher.doFinal(Hex.decode(input));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return Hex.encode(encryptedByteArray);

    }

    /**
     * 将二进制转换成16进制
     * @param buf
     * @return
     */
    private static String parseByte2HexStr(byte buf[]) {



        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < buf.length; i++) {
            String hex = Integer.toHexString(buf[i] & 0xFF);
            if (hex.length() == 1) {
                hex = '0' + hex;
            }
            sb.append(hex.toUpperCase());
        }
        return sb.toString();
        
    }
    
    /**
     * 将16进制转换为二进制
     * @param hexStr
     * @return
     */
    private static byte[] parseHexStr2Byte(String hexStr) {
        
        if (hexStr.length() < 1)
            return null;
        byte[] result = new byte[hexStr.length()/2];
        for (int i = 0;i< hexStr.length()/2; i++) {
            int high = Integer.parseInt(hexStr.substring(i*2, i*2+1), 16);
            int low = Integer.parseInt(hexStr.substring(i*2+1, i*2+2), 16);
            result[i] = (byte) (high * 16 + low);
        }
        return result;
        
    }
    
    /**
     * 解密
     * @param  input 解密的字符串
     * @param  key   解密的key
     * @return String
     */
    public String decrypt(String input, String key){

        byte[] decodeKey = Hex.decode(key);

        byte[] output = null;
        
        try{
            SecretKeySpec skey = new SecretKeySpec(decodeKey, ALGORITHM);
            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
            cipher.init(Cipher.DECRYPT_MODE, skey);
            output = cipher.doFinal(parseHexStr2Byte(input));

        }catch(Exception e){
            System.out.println(e.toString());
        }
        return new String(output);
    }
}



android/src/main/java/com/smallbuer/flutter_aes_ecb_pkcs5/FlutterAesEcbPkcs5Plugin.java
package com.smallbuer.flutter_aes_ecb_pkcs5;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.plugin.common.BinaryMessenger;

import io.flutter.embedding.engine.plugins.FlutterPlugin;

/** FlutterAesEcbPkcs5Plugin */
public class FlutterAesEcbPkcs5Plugin implements FlutterPlugin,MethodCallHandler {
  static private MethodChannel channel;

  public FlutterAesEcbPkcs5Plugin() {
    System.out.println("FlutterAesEcbPkcs5Plugin init");
  }

  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    FlutterAesEcbPkcs5Plugin plugin = new FlutterAesEcbPkcs5Plugin();
    plugin.setup(registrar.messenger());
  }

  @Override
  public void onAttachedToEngine(FlutterPluginBinding flutterPluginBinding) {
    setup(flutterPluginBinding.getBinaryMessenger());
  }

  @Override
  public void onDetachedFromEngine(FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  private void setup(
          final BinaryMessenger messenger) {
    channel = new MethodChannel(messenger, "flutter_aes_ecb_pkcs5");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    }else if (call.method.equals("generateDesKey")) {
      int length = call.argument("length");
      if (length != 0 ) {
        try {
          String encoded = new AesSecurity().generateDesKey(length);
          result.success(encoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "generateDesKey failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "generateDesKey failure.", null);
      }

    }else if(call.method.equals("encrypt")){

      String text = call.argument("input");
      String publicKey = call.argument("key");
      if (text != null && publicKey != null) {
        try {
          String encoded = new AesSecurity().encrypt(text, publicKey);
          result.success(encoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "Encrypt failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "Encrypt failure.", null);
      }
    }else if(call.method.equals("decrypt")){

      String text = call.argument("input");
      String publicKey = call.argument("key");
      if (text != null && publicKey != null) {
        try {
          String decoded = new AesSecurity().decrypt(text, publicKey);
          result.success(decoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "Decrypt failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "Decrypt failure.", null);
      }
    } else {
      result.notImplemented();
    }
  }

}


ios/Classes/FlutterAesEcbPkcs5Plugin.m
#import "FlutterAesEcbPkcs5Plugin.h"
#import <CommonCrypto/CommonCryptor.h>
#import <Foundation/Foundation.h>


@implementation FlutterAesEcbPkcs5Plugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"flutter_aes_ecb_pkcs5"
            binaryMessenger:[registrar messenger]];
  FlutterAesEcbPkcs5Plugin* instance = [[FlutterAesEcbPkcs5Plugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {

      NSDictionary *argsMap  = call.arguments;

    if ([@"getPlatformVersion" isEqualToString:call.method]) {
      result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
    } else if ([@"generateDesKey" isEqualToString:call.method]) {

        id length = [argsMap objectForKey:@"length"];
        NSLog(@"obj = %@", length);

        NSString *key = [FlutterAesEcbPkcs5Plugin randomlyGeneratedBitString:length];
        //NSLog(@"AES加密的秘钥key=%@",key);

        NSString *hexKey = [FlutterAesEcbPkcs5Plugin hexStringFromString:key];

        //NSLog(@"AES加密的秘钥hexKey=%@",hexKey);

        result(hexKey.uppercaseString);

    }else if ([@"encrypt" isEqualToString:call.method]) {

        id input = [argsMap objectForKey:@"input"];
        //NSLog(@"obj = %@", input);
        id key = [argsMap objectForKey:@"key"];
        //NSLog(@"obj = %@", key);

        NSString *encode = [FlutterAesEcbPkcs5Plugin encyptPKCS5:input WithKey:key];

        //NSLog(@"结果=%@",encode.uppercaseString);
        result(encode.uppercaseString);
    } else if ([@"decrypt" isEqualToString:call.method]) {

        id input = [argsMap objectForKey:@"input"];
        //NSLog(@"obj = %@", input);
        id key = [argsMap objectForKey:@"key"];
        //NSLog(@"obj = %@", key);

        NSString *encode = [FlutterAesEcbPkcs5Plugin decrypPKCS5:input WithKey:key];

        //NSLog(@"结果=%@",encode);
        result(encode);

    } else {
      result(FlutterMethodNotImplemented);
    }

}


+ (NSString *)randomlyGeneratedBitString:(id)length
{
    NSString *string = [[NSString alloc]init];

    NSLog(@"len->%@",length);
    int len = [length intValue] / 8 ;
    NSLog(@"len->%d",len);

    for (int i = 0; i < len; i++) {
        int number = arc4random() % 36;
        if (number < 10) {
            int figure = arc4random() % 10;
            NSString *tempString = [NSString stringWithFormat:@"%d", figure];
            string = [string stringByAppendingString:tempString];
        }else {
            int figure = (arc4random() % 26) + 97;
            char character = figure;
            NSString *tempString = [NSString stringWithFormat:@"%c", character];
            string = [string stringByAppendingString:tempString];
        }
    }
    return  string;
}


//字符串加密(16进制)
+ (NSString *)encyptPKCS5:(NSString *)plainText WithKey:(NSString *)key{

    //把string 转NSData
    // NSData* data = [plainText dataUsingEncoding:NSUTF8StringEncoding];

    //length
    // size_t plainTextBufferSize = [data length];

    // const void *vplainText = (const void *)[data bytes];

    char *hexPlainText = [FlutterAesEcbPkcs5Plugin convertHexStrToChar:plainText];
    //char *hexPlainText = malloc(16);
    //memset(hexPlainText, '0', 16);

    uint8_t *bufferPtr = NULL;
    size_t bufferPtrSize = 0;
    size_t movedBytes = 0;

    bufferPtrSize = (strlen(hexPlainText) + kCCBlockSizeAES128) & ~(kCCBlockSizeAES128 - 1);
    bufferPtr = malloc( bufferPtrSize * sizeof(uint8_t));
    memset((void *)bufferPtr, 0x0, bufferPtrSize);
    
    //解密hex加密的密钥

    char *publicKey = [FlutterAesEcbPkcs5Plugin convertHexStrToChar:key];
    //char *publicKey = malloc(16);
    //memset(publicKey, '0', 16);
    ////
    NSLog(@"vkey->len->%d",strlen(publicKey));

    //配置CCCrypt
    CCCryptorStatus ccStatus = CCCrypt(kCCEncrypt,
                                       kCCAlgorithmAES128, //3DES
                                       kCCOptionPKCS7Padding|kCCOptionECBMode, //设置模式
                                       publicKey,    //key
                                       kCCKeySizeAES256,
                                       nil,     //偏移量，这里不用，设置为nil;不用的话，必须为nil,不可以为@“”
                                    //    vplainText,
                                       hexPlainText,
                                    //    plainTextBufferSize,
                                       strlen(hexPlainText),
                                       (void *)bufferPtr,
                                       bufferPtrSize,
                                       &movedBytes);

    if (ccStatus == kCCSuccess) {
        // (NSString *) stringResult = [FlutterAesEcbPkcs5Plugin convertHexStrToChar:key];

        NSData *myData = [NSData dataWithBytes:(const char *)bufferPtr length:(NSUInteger)movedBytes];

        NSString *result = [FlutterAesEcbPkcs5Plugin hexStringForData:myData];
        return result;
        //16进制(你也可以换成base64等)
        // NSUInteger          len = [myData length];
        // char *              chars = (char *)[myData bytes];
        // NSMutableString *   hexString = [[NSMutableString alloc] init];

        // for(NSUInteger i = 0; i < len; i++ )
        //     // [hexString appendString:NSString chars[i]];
        //     [hexString appendString:[NSString stringWithFormat:@"%02x", chars[i]]];
        // return hexString;
    }

    free(bufferPtr);
    return nil;

}


//解密
+(NSString *)decrypPKCS5:(NSString *)encryptText WithKey:(NSString *)key
{

    //解密hex加密的密钥
    char *publicKey = [FlutterAesEcbPkcs5Plugin convertHexStrToChar:key];

//    NSLog(@"publicKey->publicKey->%s",publicKey);
//    NSLog(@"vkey->len->%d",strlen(publicKey));

    char *data1 = [FlutterAesEcbPkcs5Plugin convertHexStrToChar:encryptText];

    int plainTextBufferSize = [encryptText length] / 2 ;

    //NSLog(@"data1->plainTextBufferSize->%d",plainTextBufferSize);

    NSData *content=[NSData dataWithBytes:data1 length:plainTextBufferSize];

    //NSLog(@"data1->content->%@",content);

    NSUInteger dataLength = [content length];
    size_t bufferSize = dataLength + kCCBlockSizeAES128;
    void *buffer = malloc(bufferSize);

    size_t numBytesCrypted = 0;
    CCCryptorStatus cryptStatus = CCCrypt(kCCDecrypt,
                                          kCCAlgorithmAES128,
                                          kCCOptionPKCS7Padding|kCCOptionECBMode,
                                          publicKey,
                                          kCCBlockSizeAES128,
                                          NULL,
                                          [content bytes],
                                          dataLength,
                                          buffer,
                                          bufferSize,
                                          &numBytesCrypted);

    if (cryptStatus == kCCSuccess) {
        NSData *resultData = [NSData dataWithBytesNoCopy:buffer length:numBytesCrypted];

        return [[NSString alloc] initWithData:resultData encoding:NSUTF8StringEncoding];
    }
    free(buffer);
    return nil;
}

+ (char *)convertHexStrToChar:(NSString *)hexString {

    int mallocLen = [hexString length] / 2 + 1;

    char *myBuffer = (unsigned char *)malloc(mallocLen);

    memset(myBuffer,'\0',mallocLen);

    for (int i = 0; i < [hexString length] - 1; i += 2) {
        unsigned int anInt;
        NSString * hexCharStr = [hexString substringWithRange:NSMakeRange(i, 2)];
        NSScanner * scanner = [[NSScanner alloc] initWithString:hexCharStr];
        [scanner scanHexInt:&anInt];
        myBuffer[i / 2] = (char)anInt;
    }
    return myBuffer;
}

+ (NSString *)hexStringFromString:(NSString *)string{
    NSData *myD = [string dataUsingEncoding:NSUTF8StringEncoding];
    Byte *bytes = (Byte *)[myD bytes];
    //下面是Byte 转换为16进制。
    NSString *hexStr=@"";
    for(int i=0;i<[myD length];i++)
    {
        NSString *newHexStr = [NSString stringWithFormat:@"%x",bytes[i]&0xff];///16进制数
        if([newHexStr length]==1)
            hexStr = [NSString stringWithFormat:@"%@0%@",hexStr,newHexStr];
        else
            hexStr = [NSString stringWithFormat:@"%@%@",hexStr,newHexStr];
    }
    return hexStr;
}

+ (NSString *)hexStringForData:(NSData *)data
{
    if (data == nil) {
        return nil;
    }
    
    NSMutableString *hexString = [NSMutableString string];
    
    const unsigned char *p = [data bytes];
    
    for (int i=0; i < [data length]; i++) {
        [hexString appendFormat:@"%02x", *p++];
    }
    
    return hexString;
}


@end
